src/crc32c.c

```diff
@@ -0,0 +1,612 @@
+        uint32_t crc2 = 0;
+        const char *end = buf + LONG;
+        const char *buf2 = end;
+        const char *buf3 = end + LONG;
+        do {
+            crc = crc32cx(crc, unaligned_i64(buf));
+            buf += 8;
+            crc1 = crc32cx(crc1, unaligned_i64(buf2));
+            buf2 += 8;
+            crc2 = crc32cx(crc2, unaligned_i64(buf3));
+            buf3 += 8;
+        } while (buf < end);
+        crc = crc32c_shift(crc32c_long, crc) ^ crc1;
+        crc = crc32c_shift(crc32c_long, crc) ^ crc2;
+        buf += LONG * 2;
+        len -= LONG * 3;
+    }
+
+    /* do the same thing, but now on SHORT*3 blocks for the remaining data less
+     * than a LONG*3 block */
+    while (len >= SHORT * 3) {
+        uint32_t crc1 = 0;
+        uint32_t crc2 = 0;
+        const char *end = buf + SHORT;
+        const char *buf2 = end;
+        const char *buf3 = end + SHORT;
+        do {
+            crc = crc32cx(crc, unaligned_i64(buf));
+            buf += 8;
+            crc1 = crc32cx(crc1, unaligned_i64(buf2));
+            buf2 += 8;
+            crc2 = crc32cx(crc2, unaligned_i64(buf3));
+            buf3 += 8;
+        } while (buf < end);
+        crc = crc32c_shift(crc32c_short, crc) ^ crc1;
+        crc = crc32c_shift(crc32c_short, crc) ^ crc2;
+        buf += SHORT * 2;
+        len -= SHORT * 3;
+    }
+    // The same shift table can be used to compute two SHORT blocks simultaneously
+    if (len >= SHORT * 2) {
+        uint32_t crc1 = 0;
+        const char *end = buf + SHORT;
+        const char *buf2 = end;
+        do {
+            crc = crc32cx(crc, unaligned_i64(buf));
+            buf += 8;
+            crc1 = crc32cx(crc1, unaligned_i64(buf2));
+            buf2 += 8;
+        } while (buf < end);
+        crc = crc32c_shift(crc32c_short, crc) ^ crc1;
+        buf += SHORT;
+        len -= SHORT * 2;
+    }
+
+    /* compute the crc on the remaining eight-byte units less than a SHORT*2
+       block */
+    const char *end = buf + len - 8;
+    while (buf <= end) {
+        crc = crc32cx(crc, unaligned_i64(buf));
+        buf += 8;
+    }
+    if (len & 4) {
+        crc = crc32cw(crc, unaligned_i32(buf));
+        buf += 4;
+    }
+    if (len & 2) {
+        crc = crc32ch(crc, unaligned_i16(buf));
+        buf += 2;
+    }
+    if (len & 1)
+        crc = crc32cb(crc, *buf);
+    /* return a post-processed crc */
+    return ~crc;
+}
+
+// HW feature detection
+#  ifdef __ARM_FEATURE_CRC32
+// The C code is compiled with CRC32 being required. Skip runtime dispatch.
+JL_DLLEXPORT uint32_t jl_crc32c(uint32_t crc, const char *buf, size_t len)
+{
+    return crc32c_armv8(crc, buf, len);
+}
+#  else
+static crc32c_func_t crc32c_dispatch(unsigned long hwcap)
+{
+    if (hwcap & HWCAP_CRC32)
+        return crc32c_armv8;
+    return jl_crc32c_sw;
+}
+// For ifdef detection below
+#    define crc32c_dispatch() crc32c_dispatch(getauxval(AT_HWCAP))
+#    define crc32c_dispatch_ifunc "crc32c_dispatch"
```

Comment:
I don't see much of an improvement from "assume the implementation defines a function called `crc32_dispatch`" to "assume that the implementation `#defines` a symbol named `crc32c_dispatch_ifunc`".  What do you gain by allowing different names for the actual function, since only one can be defined at a time?